Source/DafnyLanguageServer/DafnyLanguageServer.cs

```diff
@@ -19,21 +21,51 @@ public static class DafnyLanguageServer {
       }
     }
 
-    public static LanguageServerOptions WithDafnyLanguageServer(this LanguageServerOptions options, IConfiguration configuration) {
+    public static LanguageServerOptions WithDafnyLanguageServer(this LanguageServerOptions options,
+      IConfiguration configuration, CancellationTokenSource cancelLanguageServer) {
       return options
         .WithDafnyLanguage(configuration)
         .WithDafnyWorkspace(configuration)
         .WithDafnyHandlers()
-        .OnInitialize(InitializeAsync)
+        .OnInitialize((server, @params, token) => InitializeAsync(server, @params, token, cancelLanguageServer))
         .OnStarted(StartedAsync);
     }
 
-    private static Task InitializeAsync(ILanguageServer server, InitializeParams request, CancellationToken cancellationToken) {
+    private static Task InitializeAsync(ILanguageServer server, InitializeParams request, CancellationToken cancelRequestToken,
+        CancellationTokenSource cancelLanguageServer) {
       var logger = server.GetRequiredService<ILogger<Program>>();
       logger.LogTrace("initializing service");
+
+      KillLanguageServerIfParentDies(logger, request, cancelLanguageServer);
+
       return Task.CompletedTask;
     }
 
+    /// <summary>
+    /// As part of the LSP spec, a language server must kill itself if its parent process dies
+    /// https://github.com/microsoft/language-server-protocol/blob/gh-pages/_specifications/specification-3-16.md?plain=1#L1713
+    /// </summary>
+    private static void KillLanguageServerIfParentDies(ILogger<Program> logger, InitializeParams request,
+      CancellationTokenSource cancelLanguageServer) {
+      if (!(request.ProcessId >= 0)) {
+        return;
+      }
+
+      void Kill() {
+        logger.LogWarning("Shutting down language server because parent process died.");
+        cancelLanguageServer.Cancel();
+      }
+
+      try {
+        var hostProcess = Process.GetProcessById((int)request.ProcessId)!;
```

Comment:
`GetProcessById` doesn't return a nullable reference hence why the null-forgiving operator `!` is unnecessary.